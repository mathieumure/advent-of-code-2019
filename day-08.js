import {readFile} from 'fs'
import {join} from 'path'
import {promisify} from 'util'
import {computeIntProgram, computeIntProgramWithState} from "./utils/intProgram";

const asyncReadFile = promisify(readFile)

const flattenDeep = (arr) => Array.isArray(arr)
    ? arr.reduce( (a, b) => a.concat(flattenDeep(b)) , [])
    : [arr]

export const parse = async path => {
    const data = await asyncReadFile(join(__dirname, path));
    const input = data.toString()
        .split('\n')[0]
        .split('')
        .map(it => parseInt(it, 10));

    return input;
}

class Layer {
    constructor(width, height) {
        this.width = width;
        this.height = height;
        this.pixels = [[]];
        this.currentLine = 0
        this.currentColumn = 0
    }

    get isLastPixelOnLine () {
        return this.currentColumn === this.width
    }

    get isLastPixelOfImage () {
        return this.currentColumn === this.width && this.currentLine === (this.height - 1)
    }

    addPixel (pixel) {
        if (this.isLastPixelOnLine) {
            this.pixels.push([])
            this.currentLine++
            this.currentColumn=0
        }
        this.pixels[this.currentLine].push(pixel)
        this.currentColumn++
    }

    getNbPixelOccurence (pixel) {
        return flattenDeep(this.pixels).filter(it => it === pixel).length
    }
}

export const parseLayers = (width, height) => input => {
    const layers = []
    let currentLayer = new Layer(width, height);

    input.forEach(pixel => {
        currentLayer.addPixel(pixel);
        if (currentLayer.isLastPixelOfImage) {
            layers.push(currentLayer)
            currentLayer = new Layer(width, height)
        }
    })

    return layers
}

export const computePart1 = items => {
    const layers = parseLayers(25,6)(items);
    const layerWithFew0 = layers.reduce((acc, layer) => {
        if (!acc) {
            return layer
        }
        if (layer.getNbPixelOccurence(0) < acc.getNbPixelOccurence(0)) {
            return layer;
        }
        return acc;
    })

    console.log(layerWithFew0)

    return layerWithFew0.getNbPixelOccurence(1) * layerWithFew0.getNbPixelOccurence(2)
}

export const computePart2 = items => {
    const layers = parseLayers(25,6)(items);
    const finalLayer = new Layer(25,6)
    for (let h = 0; h < 6; h++) {
        for (let w = 0; w < 25; w++) {
            const layerUsedForCurrentPixel = layers.find(it => it.pixels[h][w] !== 2)
            finalLayer.addPixel(layerUsedForCurrentPixel.pixels[h][w])
        }
    }
    const str = finalLayer.pixels.reduce((acc, line) => {
        const lineStr = line.map(it => {
            if (it === 0) {
                return " "
            }
            if (it === 1) {
                return "X"
            }
        }).join('')
        return acc + lineStr + '\n'
    }, '')
    console.log(str)
    return finalLayer.pixels
}

export const run = async part => {
    const data = await parse(`./data/day-08.txt`);
    const result = part === 'part1' ? computePart1(data) : computePart2(data)
    return result
}