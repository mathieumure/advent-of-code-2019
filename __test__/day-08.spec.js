import {computePart1, parseLayers, run} from '../day-08'

describe('day-O8 - part 1', () => {

    it('should find layers', () => {
        const input = [1,2,3,4,5,6,7,8,9,0,1,2]
        const layer1 = [
            [1,2,3],
            [4,5,6]
        ]
        const layer2 = [
            [7,8,9],
            [0,1,2]
        ]
        const layers = parseLayers(3,2)(input)
        expect(layers.map(it => it.pixels)).toEqual([layer1, layer2])
    })

    it('should find the layer with the fewest digit', () => {
        const input = [1,2,3,4,4,6,7,8,9,0,1,0,6, 7, 8, 3, 0, 1]
        const layers = computePart1(input)
        expect(layers).toEqual(4)
    })

    it('should run', async () => {
        const result = await run('part1')
        expect(result).toMatchSnapshot()
    })
})

describe('day-O8 - part 2', () => {
  it('should run', async () => {
      const result = await run('part2')
      expect(result).toMatchSnapshot()
  })
})
