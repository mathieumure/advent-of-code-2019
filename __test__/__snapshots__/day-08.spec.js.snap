// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`day-O8 - part 1 should run 1`] = `1474`;

exports[`day-O8 - part 2 should run 1`] = `
Layer {
  "currentColumn": 25,
  "currentLine": 5,
  "height": 6,
  "pixels": Array [
    Array [
      0,
      0,
      1,
      1,
      0,
      0,
      1,
      1,
      0,
      0,
      1,
      1,
      1,
      0,
      0,
      0,
      1,
      1,
      0,
      0,
      1,
      1,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
    ],
    Array [
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      1,
      1,
      0,
      0,
    ],
    Array [
      0,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      1,
      1,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
    ],
    Array [
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      1,
      0,
      0,
      1,
      0,
    ],
    Array [
      0,
      1,
      1,
      0,
      0,
      0,
      1,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      0,
      0,
      1,
      1,
      0,
      0,
      1,
      1,
      1,
      0,
      0,
    ],
  ],
  "width": 25,
}
`;
